v1/dictionary/handlers.js

```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+var fluid = require('infusion');
+var AdaptiveContentServices = fluid.registerNamespace('AdaptiveContentServices');
+var wd = require('word-definition');
+
+require('kettle');
+
+/* General grade for dictionary service endpoints
+ * from which all other handler grades will extend
+*/
+fluid.defaults('AdaptiveContentServices.handlers.dictionary', {
+  gradeNames: 'kettle.request.http',
+  invokers: {
+    handleRequest: 'fluid.notImplemented',
+    wikiDefinition: {
+      funcName: 'AdaptiveContentServices.handlers.dictionary.wikiDefinition',
+      args: ['{arguments}.0', '{arguments}.1']
+    },
+    checkWikiErr: {
+      funcName: 'AdaptiveContentServices.handlers.dictionary.checkWikiErr',
+      args: ['{arguments}.0']
+    }
+  }
+})
+
+/* Function to get definition from the wiktionary service
+ */
+AdaptiveContentServices.handlers.dictionary.wikiDefinition = function(lang, word) {
+  var promise = fluid.promise();
+  wd.getDef(word, lang, null, function(data) {
+    promise.resolve(data);
+  });
+  return promise;
+}
+
+/* Function to catch the errors for wiktionary service
+ */
+AdaptiveContentServices.handlers.dictionary.checkWikiErr = function(serviceResponse) {
+
+    //Check if there is an error
+    if(serviceResponse.err) {
+
+      //Word not found
+      if(serviceResponse.err == 'not found') {
+        return {
+          statusCode: 404,
+          errorMessage: 'Dictionary Service: Word not found (Wiktionary)'
+        }
+      }
+
+      //Language unsupported by the third-party service
+      else if(serviceResponse.err == 'unsupported language') {
+        return {
+          statusCode: 400,
+          errorMessage: 'Dictionary Service: Unsupported Language (Wiktionary)'
+        }
+      }
+
+      //Default return object when error hasn't been handled yet
+      else return {
+        statusCode: 500,
+        errorMessage: 'Dictionary Service: The error hasn\'t been handled yet (Wiktionary)'
+      };
+    }
+
+    //Return false if no error found
+    else return;
+}
+
+//request handler for wiktionary
+fluid.defaults('AdaptiveContentServices.handlers.dictionary.wiktionary', {
+  gradeNames: 'AdaptiveContentServices.handlers.dictionary',
+  invokers: {
+    handleRequest: {
+      funcName: 'AdaptiveContentServices.handlers.dictionary.wiktionary.getDefinition',
+      args: ['{arguments}.0', '{that}']
+    }
+  }
+});
+
+AdaptiveContentServices.handlers.dictionary.wiktionary.getDefinition = function(request, that) {
+  try {
+    var version = request.req.params.version;
```

Comment:
Since lines 84-86 will be a common need for all dictionary handlers, they should probably be extracted to a separate function.

It should become more obvious once you implement a second type of dictionary handler, but a thing to consider will be these questions:

1) What operations will be common to all the dictionary handlers?
2) What operations will be specific to particular implementations (Wiktionary, Oxford, etc)
3) How do we separate the common operations in 1 into common reusable functions, while delegating appropriately to 2?

Reply:
I was trying to do this, but couldn't figure out how to separate this into a function and return all three values (`version`, `word`, and `lang`). Although I thought of some ways but I don't think they are effective and I am sure you have something different on you mind.

Reply:
One approach would be a forwarding pattern. This could look like implementing the common functionality in a function shared between the different grades and then dispatching `version`, `word` and `lang` to service-specific functions.

In the grade definitions it might look something like this:

```
fluid.defaults('adaptiveContentServices.handlers.dictionary', {
  gradeNames: 'kettle.request.http',
  invokers: {
    // Invoker that all kettle requests are expected to have;
    // but this one simply passes to another invoker that
    // will represent the grade-specific handler
    handleRequest: {
        func: '{that}.commonDictionaryDispatcher',
        args: ["{arguments}.0", "{that}.dictionaryHandlerImpl"]
    },
    commonDictionaryDispatcher: "adaptiveContentServices.handlers.dictionary.commonDictionaryDispatcher",    
    // This is what we'll call the grade-specific handler (the "impl" is short for "implementation")
    dictionaryHandlerImpl: 'fluid.notImplemented',
    // We also expect to have code to check for an error from the service
    checkDictionaryErrorImpl: 'fluid.notImplemented'
  }
});

adaptiveContentServices.handlers.dictionary.commonDictionaryDispatcher = function (request, serviceSpecificImplementation) {
    // separate out the needed arguments from the parameter here,
    // then forward them along with the request to the 
    // service-specific function 
    serviceSpecificImplementation(request, version, word, lang);
};
```

Reply:
Here, in `serviceSpecificImplementation(request, version, word, lang)`, how to account for the second argument of the `dictionaryHandlerImpl`, i.e. `{that}`?

Reply:
You can pass it through as another positional argument. If you're using Invokers, it's very easy to do so like so:

```
fluid.defaults('adaptiveContentServices.handlers.dictionary', {
  gradeNames: 'kettle.request.http',
  invokers: {
    // Invoker that all kettle requests are expected to have;
    // but this one simply passes to another invoker that
    // will represent the grade-specific handler
    handleRequest: {
        func: '{that}.commonDictionaryDispatcher',
        args: ["{arguments}.0", "{that}.dictionaryHandlerImpl", "{that}"]
    },
    commonDictionaryDispatcher: "adaptiveContentServices.handlers.dictionary.commonDictionaryDispatcher",    
    // This is what we'll call the grade-specific handler (the "impl" is short for "implementation")
    dictionaryHandlerImpl: 'fluid.notImplemented',
    // We also expect to have code to check for an error from the service
    checkDictionaryErrorImpl: 'fluid.notImplemented'
  }
});

adaptiveContentServices.handlers.dictionary.commonDictionaryDispatcher = function (request, serviceSpecificImplementation, that) {
    // separate out the needed arguments from the parameter here,
    // then forward them along with the request to the 
    // service-specific function 
    serviceSpecificImplementation(request, version, word, lang, that);
};
```

Within the context of a grade definition, `{that}` will always refer to the current component in context, so you'll see it all over the place in grade definitions and function argument lists.